### PR TITLE
Implement a basic OpenID connect client

### PR DESCRIPTION
**Pull request by z-oo**
This is an initial implementation of OIDC using openresty-oidc plugin

### CODE REVIEW

File: deploy/mandatory.yaml [hunk 1]
```diff
@@ -209,16 +209,20 @@ spec:
         prometheus.io/scrape: "true"
     spec:
       serviceAccountName: nginx-ingress-serviceaccount
+      imagePullSecrets:
+        - name: gcr-json-key
       containers:
         - name: nginx-ingress-controller
-          image: quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.23.0
+#          image: quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.23.0
```
Comment (glukasiknuro) [inline deploy/mandatory.yaml]: Do you need this commented config?

File: deploy/mandatory.yaml [hunk 2]
```diff
@@ -261,5 +265,41 @@ spec:
...
+apiVersion: v1
+kind: Service
+metadata:
+  name: ingress-nginx
+  namespace: ingress-nginx
+  labels:
+    app.kubernetes.io/name: ingress-nginx
+    app.kubernetes.io/part-of: ingress-nginx
+#  annotations:
+#    kube-router.io/service.dsr: tunnel
```
Comment (glukasiknuro) [inline deploy/mandatory.yaml]: What's that?

File: images/nginx/rootfs/build.sh [hunk 3]
```diff
@@ -280,6 +289,16 @@ make install
 cd "$BUILD_PATH/lua-resty-cookie-0.1.0"
 make install
 
+# cd "$BUILD_PATH/lua-resty-http-0.13"
```
Comment (glukasiknuro) [inline images/nginx/rootfs/build.sh]: Can you add comment what's up with those?

File: internal/ingress/annotations/oidc/main.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+		return false
+	}
+	if c1.ClientId != c2.ClientId {
+		return false
+	}
+
+	return true
+}
+
+// Parse parses the annotations contained in the ingress
```
Comment (glukasiknuro) [inline internal/ingress/annotations/oidc/main.go]: Nit: "Parse parses" -> "Parses"

File: deploy/mandatory.yaml (hunk 2 above)
Comment (z-oo) [reply deploy/mandatory.yaml]: Removed.

It is an feature to enable direct server reply on kube router. It reduces one level of NAT when we route the reply back. For some reason it stops working recently and I don't know why. So I commented it out. Anyway, another copy of this file is under main repo's production/k8s/cluster/ingress-controller.yaml and I can revert any change here if prefered.

File: deploy/mandatory.yaml (hunk 1 above)
Comment (z-oo) [reply deploy/mandatory.yaml]: Removed

**Comment by z-oo** [comment]
Thanks for reviewing! I have removed these two commented lines.

File: images/nginx/rootfs/build.sh (hunk 3 above)
Comment (z-oo) [reply images/nginx/rootfs/build.sh]: Done.

File: internal/ingress/annotations/oidc/main.go (hunk 4 above)
Comment (z-oo) [reply internal/ingress/annotations/oidc/main.go]: Done.

**Pull request merged**